File: PrimaMateria/PrimaMateria/Classes/HelpButton.m
//
//  HelpButton.m
//  PrimaMateria
//
//  Created by Jerry Porter on 1/5/2011.
//  Copyright 2016 xTrensa. All rights reserved.
//

@interface HelpButton()
@property (NS_NONATOMIC_IOSONLY, readonly, strong) UIViewController *viewController;
@end

@implementation HelpButton

#pragma mark Private Methods

- (UIViewController *)viewController {
    UIResponder *responder = self;
    while (![responder isKindOfClass:[UIViewController class]]) {
        responder = [responder nextResponder];
        if (nil == responder) {
            break;
        }
    }
    return (UIViewController *)responder;
}

#pragma mark - Initialization Methods

#pragma mark - Misc Methods


- (void)setupColors {
    self.backgroundColor = UIColor.yellowColor;
    [self setTitleColor: UIColor.blackColor forState: UIControlStateNormal];
    self.titleEdgeInsets = UIEdgeInsetsMake(0, -5, 0, 0);
}

- (void)touchesBegan: (NSSet *)touches withEvent: (UIEvent *)event {
    NSString *label  = [self titleForState: UIControlStateDisabled];

    if(label != nil) {
        UIViewController *controller = self.viewController;
        XTRHelpBalloonViewController *content = [[XTRAppDelegate storyboard] instantiateViewControllerWithIdentifier: NSStringFromClass([XTRHelpBalloonViewController class])];
        CGRect aFrame = self.bounds;
        CGRect aRect;
        
        if([label isEqualToString: @"periodicTable"]) {
            aRect = CGRectMake(5, 5, aFrame.size.width - 15, aFrame.size.height);
        } else {
            aRect = CGRectMake(0, 0, aFrame.size.width - 5, aFrame.size.height - 5);
        }

        content.preferredContentSize = CGSizeMake(410, 338);
        content.modalPresentationStyle = UIModalPresentationPopover;
        content.popoverPresentationController.sourceRect = aRect;
        content.popoverPresentationController.sourceView = self;
        [controller presentViewController:content animated:YES completion:nil];

        [[NSNotificationCenter defaultCenter] postNotificationName: NOTIFICATION_ELEMENT_HELP_SELECTED object: label];
    }
}

#pragma mark - Memory Management Methods

@end

File: PrimaMateria/PrimaMateria/Classes/HelpButton.swift
//
//  HelpButton.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/14/16.
//  Copyright © 2016 xTrensa. All rights reserved.
//

@objc class HelpButton : UIButton, UIPopoverPresentationControllerDelegate {
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder:aDecoder)!
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        self.layer.cornerRadius = 8.0
        self.layer.masksToBounds = true
    }
    
    // MARK: - Internal Methods
    
    func viewController() -> UIViewController {
        var responder: UIResponder? = self
        
        while !(responder!.isKindOfClass(UIViewController.classForCoder())) {
            responder = responder!.nextResponder()!
            if responder == nil {
                break
            }
        }
        return responder as! UIViewController
    }
    
    func setupColors() {
        self.backgroundColor = UIColor.yellowColor()
        self.setTitleColor(UIColor.blackColor(), forState: UIControlState.Normal)
        self.titleEdgeInsets = UIEdgeInsetsMake(0, -5, 0, 0);
    }
    
    // MARK: - Misc Methods
    
    override func touchesBegan(touches: Set<UITouch>, withEvent event: UIEvent?) {
        let label : String? = self.titleForState(UIControlState.Disabled)!
        
        if(label != nil) {
            let controller : UIViewController = self.viewController();
            let content : XTRHelpBalloonViewController = XTRAppDelegate.storyboard().instantiateViewControllerWithIdentifier("XTRHelpBalloonViewController") as! XTRHelpBalloonViewController
            let aFrame : CGRect = self.bounds
            let aRect : CGRect?
            
            if(label == "periodicTable") {
                aRect = CGRectMake(5, 5, aFrame.size.width - 15, aFrame.size.height);
            } else {
                aRect = CGRectMake(0, 0, aFrame.size.width - 5, aFrame.size.height - 5);
            }
            
            content.preferredContentSize = CGSizeMake(410, 338);
            content.modalPresentationStyle = UIModalPresentationStyle.Popover;
            content.popoverPresentationController!.sourceRect = aRect!;
            content.popoverPresentationController!.sourceView = self;
            controller.presentViewController(content, animated:true, completion:nil)
            content.popoverPresentationController!.backgroundColor = XTRColorFactory.popupArrowColor()
            NSNotificationCenter.defaultCenter().postNotificationName(NOTIFICATION_ELEMENT_HELP_SELECTED, object: label)
        }
    }
}
